File: frontend-desktop/DTO/Comment.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DTO.Utils;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace DTO
{
    public class Comment
    {
        [JsonProperty("_id")]
        public string Id { get; set; }

        [JsonProperty("replyFor")]
        public string ReplyFor { get; set; }

        [JsonProperty("text")]
        public string Text { get; set; }

        [JsonProperty("voice")]
        public string Voice { get; set; }

        [JsonProperty("images")]
        public List<ImageModel> Images { get; set; } = new List<ImageModel>();

        [JsonProperty("createdAt")]
        public DateTime CreatedAt { get; set; }

        [JsonProperty("updatedAt")]
        public DateTime UpdatedAt { get; set; }

        [JsonProperty("likedBy")]
        public List<string> LikedBy { get; set; } = new List<string>();

        [JsonProperty("status")]
        public string Status { get; set; }

        [JsonProperty("replies")]
        public List<Comment> Replies { get; set; } = new List<Comment>();

        [JsonProperty("createdBy")]
        public UserSummary Author { get; set; }

        public Dictionary<string, object> ToDictionary()
        {
            return new Dictionary<string, object>
            {
                { "_id", Id },
                { "replyFor", ReplyFor },
                { "text", Text },
                { "voice", Voice },
                { "images", Images.ConvertAll(img => img.ToDictionary()) },
                { "createdAt", CreatedAt },
                { "updatedAt", UpdatedAt },
                { "likedBy", LikedBy },
                { "status", Status },
                { "replies", Replies.ConvertAll(reply => reply.ToDictionary()) },
                { "createdBy", Author?.ToDictionary() }
            };
        }

        public static Comment FromDictionary(Dictionary<string, object> data)
        {
            var comment = new Comment
            {
                Id = DictionaryConverter.GetValueOrDefault(data, "_id", string.Empty),
                ReplyFor = DictionaryConverter.GetValueOrDefault(data, "replyFor", string.Empty),
                Text = DictionaryConverter.GetValueOrDefault(data, "text", string.Empty),
                Voice = DictionaryConverter.GetValueOrDefault(data, "voice", string.Empty),
                Images = DictionaryConverter.ParseImageModelList(data, "images"),
                CreatedAt = DictionaryConverter.ParseDateTime(data, "createdAt", DateTime.MinValue),
                UpdatedAt = DictionaryConverter.ParseDateTime(data, "updatedAt", DateTime.MinValue),
                LikedBy = DictionaryConverter.ParseStringList(data, "likedBy"),
                Status = DictionaryConverter.GetValueOrDefault(data, "status", string.Empty),
                Replies = DictionaryConverter.ParseCommentList(data, "replies"),
                Author = DictionaryConverter.ParseUserSummary(data, "createdBy"),
            };

            return comment;
        }

    }
    
}

File: backend/internal/models/comment.go
package models

import (
	"multiaura/pkg/utils"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Comment model
type Comment struct {
	ID            primitive.ObjectID `bson:"_id,omitempty" json:"_id,omitempty" form:"_id,omitempty"`
	ReplyFor      string             `bson:"replyFor,omitempty" json:"replyFor,omitempty" form:"replyFor,omitempty"`
	Text          string             `bson:"text" json:"text" form:"text"`
	Voice         string             `bson:"voice,omitempty" json:"voice,omitempty" form:"voice,omitempty"`
	Images        []Image            `bson:"images,omitempty" json:"images,omitempty" form:"images,omitempty"`
	CreatedAt     time.Time          `bson:"createdAt" json:"createdAt" form:"createdAt"`
	UpdatedAt     time.Time          `bson:"updatedAt" json:"updatedAt" form:"updatedAt"`
	LikedBy       []string           `bson:"likedBy" json:"likedBy" form:"likedBy"`
	Status        string             `bson:"status" json:"status" form:"status"`
	Replies       []Comment          `bson:"replies,omitempty" json:"replies,omitempty" form:"replies,omitempty"`
	CreatedBy     UserSummary        `bson:"createdBy" json:"createdBy" form:"createdBy"`
	ToxicityScore float64            `bson:"toxicityScore,omitempty" json:"toxicityScore,omitempty" form:"toxicityScore,omitempty"`
}

// CreateCommentRequest represents the structure of the request to create a new comment
type CreateCommentRequest struct {
	Text     string `bson:"text" json:"text" form:"text"`
	ReplyFor string `bson:"replyFor,omitempty" json:"replyFor,omitempty" form:"replyFor,omitempty"`
	// Voice string `bson:"voice,omitempty" json:"voice,omitempty" form:"voice,omitempty"`
}

func (c *Comment) ToMap() map[string]interface{} {
	images := make([]map[string]interface{}, len(c.Images))
	for i, img := range c.Images {
		images[i] = map[string]interface{}{
			"url": img.URL,
			"_id": img.ID,
		}
	}

	replies := make([]map[string]interface{}, len(c.Replies))
	for i, reply := range c.Replies {
		replies[i] = reply.ToMap()
	}

	return map[string]interface{}{
		"_id":           c.ID,
		"replyFor":      c.ReplyFor,
		"text":          c.Text,
		"voice":         c.Voice,
		"images":        images,
		"createdAt":     c.CreatedAt,
		"updatedAt":     c.UpdatedAt,
		"likedBy":       c.LikedBy,
		"status":        c.Status,
		"replies":       replies,
		"createdBy":     c.CreatedBy.ToMap(),
		"toxicityScore": c.ToxicityScore,
	}
}

func (c *Comment) FromMap(data map[string]interface{}) (*Comment, error) {
	imageData := utils.GetArrayMap(data, "images")
	images := make([]Image, len(imageData))
	for i, img := range imageData {
		images[i] = Image{
			URL: utils.GetString(img, "url"),
			ID:  utils.GetObjectID(img, "_id"),
		}
	}

	replyData := utils.GetArrayMap(data, "replies")
	replies := make([]Comment, len(replyData))
	for i, reply := range replyData {
		replyMap := reply
		rep, err := new(Comment).FromMap(replyMap)
		if err != nil {
			return nil, err
		}
		replies[i] = *rep
	}

	createdByData := utils.GetMap(data, "createdBy")
	createdBy, err := new(UserSummary).FromMap(createdByData)
	if err != nil {
		return nil, err
	}

	return &Comment{
		ID:            utils.GetObjectID(data, "_id"),
		ReplyFor:      utils.GetString(data, "replyFor"),
		Text:          utils.GetString(data, "text"),
		Voice:         utils.GetString(data, "voice"),
		Images:        images,
		CreatedAt:     utils.GetTime(data, "createdAt"),
		UpdatedAt:     utils.GetTime(data, "updatedAt"),
		LikedBy:       utils.GetStringArrayFromPrimitiveAMap(data, "likedBy"),
		Status:        utils.GetString(data, "status"),
		Replies:       replies,
		CreatedBy:     *createdBy,
		ToxicityScore: utils.GetFloat64(data, "toxicityScore"),
	}, nil
}
